### PR TITLE
Optimized unsplit method for quadtree; Target selection

### PR DESCRIPTION
**Pull request by Amyantis**
Quadtree:
- split
- unsplit
- get nodes at position
- draw zones and debug node shape

Game:
- remodeling entity, character and player (need discussion)
- select player target on click (display its bounds and its name)

Quadtree (split/unsplit/get nodes at position/debug); target

### CODE REVIEW

**Comment by Shiroy** [comment]
I have this error. You must have forgotten to commit a file:
`Failed to load font "assets/font/FreeSans.ttf" (failed to create the font face)`

**Comment by Shiroy** [comment]
Can you make the quadtree to be hidden on start-up and add a shortcut to display it. 
Or add a command line argument to explicitly display it.

File: engine/src/QuadTree.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        isSplit(false) {
+
+    shape.setPosition(x, y);
+
+    sf::Vector2f size(width, height);
+    shape.setSize(size);
+}
+
+QuadTree::~QuadTree() {
+    delete northEast;
```
Comment (Shiroy) [inline engine/src/QuadTree.cpp]: You should use `std::unique_ptr` instead and make the quadtree non-copyable. That would prevent a lot of future problem. Moreover, we will often use this structure to optimize our collision detection check. We cannot take the risk to have memory problem when they can be avoided.

( Have a look here if you're a bit lost : http://www.umich.edu/~eecs381/handouts/C++11_smart_ptrs.pdf )

File: engine/src/QuadTree.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        if (node->getGlobalBounds().contains(x, y)) nodesAt.push_back(node);
+    }
+    return nodesAt;
+}
+
+std::list<sf::Sprite *> QuadTree::getNodesAt(const sf::Vector2f pos) {
+    return getNodesAt(pos.x, pos.y);
+}
+
+void QuadTree::draw(sf::RenderTarget &canvas) {
```
Comment (Shiroy) [inline engine/src/QuadTree.cpp]: Make your quad tree derive of [`sf::Drawable`](http://www.sfml-dev.org/documentation/2.0-fr/classsf_1_1Drawable.php) instead. That's the best practice to draw stuff on screen, even it is for debugging.

Then you'll be able to do `mWindow->draw(quadtree);`

File: engine/src/QuadTree.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    std::list<sf::Sprite *> nodes;
+
+    bool isSplit;
+
+    QuadTree *northWest;
+    QuadTree *northEast;
+    QuadTree *southWest;
+    QuadTree *southEast;
+
```
Comment (Shiroy) [inline engine/src/QuadTree.h:29]: See the remark on the `unique_ptr` above

File: client/src/Game.cpp [hunk 4]
```diff
@@ -47,20 +71,15 @@ void Game::update(sf::Time deltaTime) {
...
-        movement.x -= 1;
-    if (playerCommands.isMIsMovingRight())
-        movement.x += 1;
-
-    if(movement != sf::Vector2f(0.0f, 0.0f)) {
-        normalize(movement);
-        movement *= player.getSpeed();
-        player.move(movement * deltaTime.asSeconds());
-    }
+    if (playerCommands.isMovingUp())
```
Comment (Shiroy) [inline client/src/Game.cpp:75]: This code should be moved to `Player` or `PlayerCommands`

Comment (Amyantis) [inline client/src/Game.cpp:75]: Ok, I will add a Player in PlayerCommands in order to move this code in PlayerCommands.

**Comment by Amyantis** [comment]
- Pointers replaced with smartpointers
- FreeSans.ttf commited
- Draw on shortcut for next update

File: client/CMakeLists.txt [hunk 5]
```diff
@@ -2,7 +2,7 @@ set(source_client
         src/main.cpp
         src/Game.cpp src/Game.h
         src/Player.cpp src/Player.h
-        src/PlayerCommands.cpp src/PlayerCommands.h)
+        src/PlayerCommands.cpp src/PlayerCommands.h ../engine/src/Character.cpp ../engine/src/Character.h)
```
Comment (Shiroy) [inline client/CMakeLists.txt:5]: This is the wrong `CMakeLists.txt` for engine files

File: engine/src/QuadTree.cpp [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    sf::Vector2f size(width, height);
+    shape.setSize(size);
+
+    shape.setFillColor(sf::Color::Transparent);
+    shape.setOutlineColor(sf::Color::Cyan);
+    shape.setOutlineThickness(2);
+}
+
+QuadTree::~QuadTree() {
+    northEast.reset(nullptr);
```
Comment (Shiroy) [inline engine/src/QuadTree.cpp:30]: You don't need to do that. The object will be destroyed automatically. The main advantage of using smart pointers is that you don't manage the memory manually

**Pull request closed without merging**